capsules/extra/src/nonvolatile_storage_driver.rs

```diff
@@ -38,6 +77,44 @@
 //! +-----------------------------------------------------------------+
 //! ```
 //!
+//! Example nonvolatile storage layout:
+//!
+//! ```text
+//!     ╒════════ ← Start of kernel region
+//!     │
+//!     │
+//!     │
+//!     ╞════════ ← Start of userspace region
+//!     │ magic header (u32)
+//!     ├──────── ← Start of App 1's region header
+//!     │ App 1's ShortID (u32)  
+//!     │ region 1 length (usize)
+//!     ├──────── ← Start of App 1's Region          ═╗
```

Comment:
It maybe makes sense to include the `magic header` in each of the app's region headers? That has a few potential benefits:
 - Currently there is a hidden, second magic value (i.e. 0) which is used to demarcate the end of the list
    - And if `0` is missing for whatever reason, (e.g. a power failure when writing App 2's region header, but before it could write the "no App 3" header) this will do bad/buggy things.
 - The magic value could be used as a form of versioning for the region header
    - This may be more complex than helpful, but it could also be a kind of validity check for the header. i.e., `const MAGIC_HEADER_V1 = 0x....`, and you write to the region `magic_value = shortID ^ length ^ MAGIC_HEADER_V1` or similar

Reply:
I'm in favor of adding a `magic_header` per region especially since it'll no longer require the terminating `0` magic value at the end (which I found kind of clunky).

Using it for versioning would be convenient as well since you would have versioning on a region by region basis. 

So based on the idea you mentioned, when the capsule reads the header it'll take `magic_value`, XOR it with `shortID` and `length`, and then check if it matches any of the supported `MAGIC_HEADER` versions (eg: `MAGIC_HEADER_V1 or MAGIC_HEADER_V2`) ?